test: add integration tests for GcsUploader using testcontainers library

Herraj opened the pull request:
## References

JIRA: [DISCO-2767](https://mozilla-hub.atlassian.net/browse/DISCO-2767)

## Description
Add integration tests for `GcsUploader` using `testcontainers` library

We had to switch to using `machine` executors for running the `integration-tests` check since the `testcontainers` test spin up a docker container at run time. This wasn't working if we were using the docker container to run those tests in. Surprisingly, setting up a python environment on a bare bones ubuntu VM and running the tests seems to be slightly faster than using a pre-built python docker image

In the `project.toml` file, you'll see that the new dev dependencies are added under `tool.poetry.group.dev.dependencies` and not `tool.poetry.dev.dependencies`. This is because the latter has been deprecated. Poetry wants us to use `...group...` now. Can make this change for the rest of the dev dependencies in the next PR.

## PR Review Checklist

- [x] This PR conforms to the [Contribution Guidelines](https://github.com/mozilla-services/merino-py/blob/main/CONTRIBUTING.md)
- [x] The PR title starts with the JIRA issue reference, format `[DISCO-####]`, and has the same title (if applicable)
- [ ] `[load test: (abort|warn)]` keywords are applied (if applicable)
- [x] [Documentation](https://github.com/mozilla-services/merino-py/tree/main/docs) has been updated (if applicable)
- [ ] [Functional and performance test](https://github.com/mozilla-services/merino-py/blob/main/docs/dev/testing.md) coverage has been expanded and maintained (if applicable)


[DISCO-2767]: https://mozilla-hub.atlassian.net/browse/DISCO-2767?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ncloudioj commented on tests/integration/content_handler/test_gcp_uploader.py:
```diff
@@ -0,0 +1,120 @@
+    """Create and return a docker container for google cloud storage entities. Tear it down
+    after all the tests have finished running
+    """
+    print("Starting up GCS storage container")
```
Nit: we normally prefer logging instead of printing. Something like:

```py
import logging

logger = logging.getLogger(__name__)

logger.info("....")
```

ncloudioj reviewed (changes requested):
Thanks for taking a crack at it! This looks good!

I dropped a few nits, also, could you please update the [integration test doc](https://github.com/mozilla-services/merino-py/blob/main/docs/testing/integration-tests.md) to note for this change as now Testcontainers and Docker are required in order to run integration tests locally. Thanks!

Trinaa commented on tests/integration/content_handler/test_gcp_uploader.py:
```diff
@@ -0,0 +1,120 @@
+    yield uploader
+
+
+def test_upload_image(
+    gcs_storage_container, gcs_storage_client, gcs_storage_bucket, gcp_uploader
+):
```
The gcs_storage_container and gcs_storage_client are unused in this test and can be removed

```suggestion
def test_upload_image(gcs_storage_bucket, gcp_uploader):
```

Unless you take my gcs_uploader suggestion, in which case the gcs_storage_container can be removed.

```suggestion
test_upload_image(gcs_storage_client, gcs_storage_bucket):
```

Trinaa commented on tests/integration/content_handler/test_gcp_uploader.py:
```diff
@@ -0,0 +1,120 @@
+    assert download_blob == image_content
+
+
+def test_get_most_recent_file(
+    gcs_storage_container, gcs_storage_client, gcs_storage_bucket, gcp_uploader
+):
```
The gcs_storage_container, gcs_storage_client and gcs_storage_bucket are unused in this test and can be removed

```suggestion
def test_get_most_recent_file(gcp_uploader):
```

Unless you take my gcs_uploader suggestion, in which case the gcs_storage_container can be removed.

```suggestion
test_get_most_recent_file(gcs_storage_client, gcs_storage_bucket):
```

Trinaa commented on tests/integration/content_handler/test_gcp_uploader.py:
```diff
@@ -0,0 +1,120 @@
+    bucket.delete(force=True)
+
+
+@pytest.fixture(scope="function")
+def gcp_uploader(
+    gcs_storage_client, gcs_storage_bucket
+) -> Generator[GcsUploader, None, None]:
+    """Return a GcsUploader instance for each test"""
+    uploader = GcsUploader(
+        destination_gcp_project=gcs_storage_client.project,
+        destination_bucket_name=gcs_storage_bucket.name,
+        destination_cdn_hostname="test_cdn_hostname",
+    )
+    yield uploader
```
In the interest of better comprehension, I suggest to remove this fixture and instantiate the GcsUploader, the test target, in the tests themselves.

Trinaa reviewed (commented):
From looking at the coverage report, I see opportunity to add more coverage both in the GcsUploader and in the navigation suggestion commands that use the GcsUploader.

<img width="795" alt="image" src="https://github.com/mozilla-services/merino-py/assets/3422688/5069ab9e-fcba-42d7-8318-36df31dfecf4">

<img width="676" alt="image" src="https://github.com/mozilla-services/merino-py/assets/3422688/c3f1ea8f-6014-4475-9e38-614f2c40b27a">


Is that planned?

Put another way, so far these tests feel unit like, but this is the integration layer, can we exercise the GcsUploader from the nav suggestions?

Herraj commented:
> Thanks for taking a crack at it! This looks good!
> 
> I dropped a few nits, also, could you please update the [integration test doc](https://github.com/mozilla-services/merino-py/blob/main/docs/testing/integration-tests.md) to note for this change as now Testcontainers and Docker are required in order to run integration tests locally. Thanks!

Thanks, I updated the doc. Let me know if you find the wording and formatting correct!

Herraj commented:
> Thank you @Herraj for getting this project underway! I'm so excited to see testcontainers in action! 🚀 🌔
> 
> From looking at the coverage report, I see opportunity to add more coverage both in the GcsUploader and in the navigation suggestion commands that use the GcsUploader.
> <img alt="image" width="795" src="https://private-user-images.githubusercontent.com/3422688/319637882-5069ab9e-fcba-42d7-8318-36df31dfecf4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIzNDM5NTIsIm5iZiI6MTcxMjM0MzY1MiwicGF0aCI6Ii8zNDIyNjg4LzMxOTYzNzg4Mi01MDY5YWI5ZS1mY2JhLTQyZDctODMxOC0zNmRmMzFkZmVjZjQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDVUMTkwMDUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTk2MTJkZjA3ZDMyM2NiYzg2NzZiMGU4ZTI2NWFjZjA1YzY5NDRiYzM1YzlmMDdkMDNkNDc3MzAzNWFhYzlhOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.UGTIRq9ZgsgmS_8cNNjw1rIlS5S69EZUgUBN3Lap_kM"> <img alt="image" width="676" src="https://private-user-images.githubusercontent.com/3422688/319639705-c3f1ea8f-6014-4475-9e38-614f2c40b27a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIzNDM5NTIsIm5iZiI6MTcxMjM0MzY1MiwicGF0aCI6Ii8zNDIyNjg4LzMxOTYzOTcwNS1jM2YxZWE4Zi02MDE0LTQ0NzUtOWUzOC02MTRmMmM0MGIyN2EucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDVUMTkwMDUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2FhNzEyZDJhZGMwMTg2YzA4MzJkZDM2NTYyNjNkODMyZWU5NWYxMmU2MTgwZjcyODhmOTJiNmY2OTYxMTk5NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.DaJsvx9x9LYryTNOaMIboLQo5S4VF53Qpqq0UwOYj1I">
> 
> Is that planned?
> 
> Put another way, so far these tests feel unit like, but this is the integration layer, can we exercise the GcsUploader from the nav suggestions?

Thanks for the feedback. I agree, the tests were looking like unit tests. I refactored the whole thing and pulled it one level above in the `DomainMetadataUploader` class where these `GcsUploader` methods are used. Let me know what you think about the changes!

ncloudioj reviewed (approved):
First step with Testcontainers, fascinating! Thanks again!

Trinaa reviewed (approved):
Thanks @Herraj !!! 🚀
